erddapy/objects/objects.py

```diff
@@ -0,0 +1,189 @@
+        self._constraints = constraints
+        self._meta = None
+
+    @property
+    def connection(self) -> ERDDAPConnection:
+        """Access private ._connection variable."""
+        return self._connection
+
+    @connection.setter
+    def connection(self, value: str | ERDDAPConnection):
+        """Set private ._connection variable."""
+        self._connection = ERDDAPConnection(ERDDAPConnection.to_string(value))
+
+    def get(self, file_type: str) -> StrLike:
+        """Request data using underlying connection."""
+        return self.connection.get(file_type)
+
+    def open(self, file_type: str) -> FilePath:
+        """Download and open dataset using underlying connection."""
+        return self.connection.open(file_type)
+
+    def get_meta(self):
+        """Request dataset metadata from the server."""
+        self._meta = None
+
+    @property
+    def meta(self):
+        """Access private ._meta attribute. Request metadata if ._meta is empty."""
+        return self.get_meta() if (self._meta is None) else self._meta
+
+    @property
+    def variables(self):
+        """Access private ._variables attribute."""
+        return self._variables
+
+    @property
+    def constraints(self):
+        """Access private ._constraints attribute."""
+        return self._constraints
+
+    def url_segment(self, file_type: str) -> str:
+        """Return URL segment without the base URL (the portion after 'https://server.com/erddap/')."""
+        pass
+
+    def url(self, file_type: str) -> str:
+        """
+        Return a URL constructed using the underlying ERDDAPConnection.
+
+        The URL will contain information regarding the base class server info, the dataset ID,
+        access method (tabledap/griddap), file type, variables, and constraints.
+
+        This allows ERDDAPDataset subclasses to be used as more opinionated URL constructors while still
+        not tying users to a specific IO method.
+
+        Not guaranteed to capture all the specifics of formatting a request, such as if a server requires
+        specific auth or headers.
+        """
+        pass
+
+    def to_dataset(self):
+        """Open the dataset as xarray dataset by downloading a subset NetCDF."""
+        pass
+
+    def opendap_dataset(self):
+        """Open the full dataset in xarray via OpenDAP."""
+        pass
+
+
+class TableDataset(ERDDAPDataset):
+    """Subclass of ERDDAPDataset specific to TableDAP datasets."""
+
+    def to_dataframe(self):
+        """Open the dataset as a Pandas DataFrame."""
+
+
+class GridDataset(ERDDAPDataset):
+    """Subclass of ERDDAPDataset specific to GridDAP datasets."""
+
+    pass
+
+
+class ERDDAPServer:
+    """Instance of an ERDDAP server, with support to ERDDAP's native functionalities."""
+
+    def __init__(self, connection: str | ERDDAPConnection):
+        """Initialize instance of ERDDAPServer."""
+        self._connection = ERDDAPConnection(ERDDAPConnection.to_string(connection))
```

Comment:
```suggestion
    def __init__(self, url: str, connection: ERDDAPConnection | None):
        """Initialize instance of ERDDAPServer."""
        if "http" in url:
            self.url = url
        else:
            # get URL from dict of ERDDAP servers
        self._connection = connection or ERDDAPConnection()
```

If the connection doesn't include server info, then the connection can be simplified.

We could also support initializing servers based on whats in the [`servers.server_list`](https://github.com/ioos/erddapy/blob/6057872825aa29423e31c849a3c921fe25474e21/erddapy/servers/servers.py#L10) as we currently do.